teps/0032-tekton-notifications.md

```diff
@@ -0,0 +1,167 @@
+in the main pipeline.
+
+Tekton users would highly benefit from native notification support.
+
+### Goals
+
+The main goal of the TEP is to design and implement a notification system that
+is both easy to use and flexible, for delivering notification targeted to humans.
+
+Notifications are easy to use: enabling and configuring default notifications
+to get something up and running (notifying on every event of every Pipeline/Task)
+requires only few lines of configuration.
+
+Notifications are flexible: users are able to configure the precise Pipelines/Tasks
+and events that notifications are sent for. Users can also modify the configuration
+for their chosen notification system.
+
+A secondary goal is to make sure that notification can benefit from our growing
+catalog of tasks, several of which have been written to deliver notification through
+different platforms.
+
+### Non-Goals
+
+- Build a machine to machine communication protocols
+- Build pre/post execution hook for steps, tasks or pipelines
+- Build a catalog of integration with different notification, chat or email systems
+- Build a system to send large amounts of data produced by a task or pipeline
+- Write a controller for cloud events (even if that could be part of the solution).
+  See [tektoncd/pipeline#2944](https://github.com/tektoncd/pipeline/issues/2944) for
+  further context.
+
+### Use Cases (optional)
+
+- Notify individuals or groups about the failure of a pipeline
+- Notify individuals or groups about the completion of a long running task
+- Notify individuals or groups about the start or a pipeline that was triggered as
+  a consequence of an automated event (cron, pull request, manual approval)
+
+## Requirements
+
+This section is a stub and subject to change.
+
+Once notifications are implemented, we do not want to force Tekton users to modify
+all their existing tasks or pipelines to enable notifications for them:
+
+- Allow using catalog tasks as is
+- Allow notifications for pipelines with no change to the pipeline itself
+
+For installations that run large number of task and pipelines, the configuration
+of notification should be designed to be scalable and easy to maintain.
+
+- Allow selecting pipelines and tasks to be notified about via selectors
+- Allow selecting events to be notified about
+- Allow pipeline and task authors to re-use of notification "rules"
+
+## Proposal
+
+This section is a stub and subject to change.
+
+The proposal is to introduce notification policies and templates.
+
+Notification policies define condition under which we want to trigger a notification;
+they would allow users to:
+
+- filter sources such using rules (using CEL?)
+  - kind: task, pipeline
+  - metadata: name, labels, namespace, annotations(?)
+- filter events such as start, run, success, fail (using CEL or syntax sugar on top)
+
+Notification templates define how a notification is sent. They would allow user to:
```

Comment:
+100 for templating. Do we want to have a standard templating language like Mustache? Or do we want to leave it up to the notifier implementation?

Also, we should allow for bindings in addition to templates. This would allow us to have notifications for specific users. like a PipelineRun's triggering commit author.

Reply:
My plan here was to leverage trigger templates and trigger bindings, i.e. a notification template could include references to a trigger template and bindings, or it could include the information required to provision them.
The trigger template would typically use tasks from the catalog.

We can already achieve notifications using pipelines and triggers todays, but it takes a lot of plumbing things together, and also it's not easily reusable.

Reply:
Reusing trigger templating and bindings sounds good to me 👍  (cc @dibyom ).